### DISCUSSION TITLE
de.Variable warmup restore

### DISCUSSION THREAD

**Discussion by dakabang** [tensorflow/recommenders-addons#213] [Ideas]
# de.Variable warmup restore

### tfra.WarmupCheckpoint

In tf2.x we use `tf.train.Checkpoint` for saving/restoring model, but I found it a little hard when I use `tf.train.Checkpoint` to restore part of the model to perform transfer learning. In recommendation system, it's a common trick to speed up model convergence using embedding warmup, and the dense part of the model is trained from scratch, hence partial restore. The demonstrate the API design for partial restore are as follows.

API design
```python

## Create model
class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.user_embeddings = tfra.dynamic_embedding.get_variable(...)
        ....
            
model = Model()
checkpoint = tf.train.Checkpoint(model)
warmup_checkpoint = tfra.WarmupCheckpoint(checkpoint)

## Only variable in var_list is restored from checkpoint
warmup_checkpoint.restore(
    restore_path, var_list=[model.user_embeddings])
```

Possible implementation(pseudo code)
```python
def WarmupCheckpoint(self):
	assert isinstance(checkpoint, tf.train.Checkpoint)
	self.restore = _restore

	def _restore(save_path, assign_map):
		vars = ... get all variable ....
                 for var in vars:
                     if var in var_list:
                         ... restore var from checkpoint ...
		...
```

**Comment by Lifann** [comment]
By assuming user_embeddings is already a list:
```python
warmup_checkpoint.restore(
    restore_path, var_list=model.user_embeddings)
```

**Reply by dakabang** [reply]
My mistake, user_embeddings is a de.Variable, fix it already.

**Comment by rhdong** [comment]
Hi @dakabang ,Thank you for your work!
Basically, I recommends you make the API easy to be understood, and if possible, you can refer to the native TF API design.
I have some questions and suggestions:

1. The 'warm start' would be more suitable for the API naming, you know, the 'warm up' is usually used as training small part dataset before formal training.
2. We'd better ensure the `tfra.WarmupCheckpoint` can work with estimator API and work on graph mode, some unit tests?
3. You can refer to the TF estimator API [tf.estimator.WarmStartSettings](https://www.tensorflow.org/api_docs/python/tf/estimator/WarmStartSettings) for naming, inputs design, especially, it'll be better to provide a strings filter mechanism for **var to warm start** based on **regular expression**. if you want to do that you can refer to [here](https://github.com/tensorflow/tensorflow/blob/abc1f2c18cc4f722bc29ffb9ce699f7cd4ed542c/tensorflow/python/training/warm_starting_util.py#L424). BTW, I don't think the `var_name_to_vocab_info` and `var_name_to_prev_var_name` are necessary here, we can make thing simple.

**Comment by rhdong** [comment]
I have talked with @Lifann , we found the keras has no warm start mechanism, `load_weights` cann't restore part of varibles, so we trend to follow estimator design,
```python
model = Model()
checkpoint = tf.train.Checkpoint(model)
restore_emb = tfra.dynamic_embedding.warm_start(ckpt_to_initialize_from='/path/to/ckpt/', vars_to_warm_start='*user_id*')
```
